File: engine-data/src/main/java/com/sos/scheduler/engine/data/log/LogEvent.java
package com.sos.scheduler.engine.data.log;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.sos.scheduler.engine.data.event.AbstractEvent;
import com.sos.scheduler.engine.data.message.MessageCode;
import scala.Option;

import java.util.regex.Matcher;
import java.util.regex.Pattern;
import javax.annotation.Nullable;

public class LogEvent extends AbstractEvent {
    private static final Pattern codePattern = Pattern.compile("([A-Z]+(-[0-9A-Z]+)+)( .*)?", Pattern.DOTALL);

    private final SchedulerLogLevel level;
    private final String line;

    protected LogEvent(SchedulerLogLevel level, String message) {
        this.level = level;
        this.line = message;
    }

    public final Option<MessageCode> codeOption() {
        String code = getCodeOrNull();
        return code == null? Option.<MessageCode>apply(null) : Option.apply(new MessageCode(code));
    }

    @JsonIgnore @Nullable
    public final String getCodeOrNull() {
        return messageCodeFromLineOrNull(line);
    }

    public final SchedulerLogLevel level() {
        return level;
    }

    @JsonGetter
    public final String message() {
        return line;
    }

    @Override public String toString() {
        return super.toString() +", line="+ line;
    }

    @Nullable static String messageCodeFromLineOrNull(String line) {
        Matcher m = codePattern.matcher(line);
        return m.matches()? m.group(1) : null;
    }

    public static LogEvent of(SchedulerLogLevel level, String line) {
        switch (level) {
            case info: return new InfoLogEvent(line);
            case warning: return new WarningLogEvent(line);
            case error: return new ErrorLogEvent(line);
            default: return new LogEvent(level, line);
        }
    }
}

File: engine-data/src/main/scala/com/sos/scheduler/engine/data/log/LogEvent.scala
package com.sos.scheduler.engine.data.log

import com.sos.scheduler.engine.base.sprayjson.typed.{SimpleTypedJsonFormat, TypedJsonFormat}
import com.sos.scheduler.engine.data.event.NoKeyEvent
import com.sos.scheduler.engine.data.log.LogEvent._
import com.sos.scheduler.engine.data.message.MessageCode
import scala.PartialFunction.condOpt
import spray.json.DefaultJsonProtocol._
import spray.json._

sealed trait LogEvent extends NoKeyEvent {

  def level: SchedulerLogLevel
  def message: String

  final def codeOption: Option[MessageCode] = messageToCode(message)
}

sealed trait InfoOrHigherLogged extends LogEvent

final case class InfoLogged(message: String) extends InfoOrHigherLogged {
  def level = SchedulerLogLevel.info
}

sealed trait WarningOrHigherLogged extends InfoOrHigherLogged

final case class WarningLogged(message: String) extends WarningOrHigherLogged {
  def level = SchedulerLogLevel.warning
}

final case class ErrorLogged(message: String) extends WarningOrHigherLogged {
  def level = SchedulerLogLevel.error
}

private case class OtherLevelLogged(level: SchedulerLogLevel, message: String)
extends LogEvent

object LogEvent {
  private val CodeRegex = "^([A-Z]+(-[0-9A-Z]+)+)".r.unanchored

  def messageToCode(message: String): Option[MessageCode] =
    condOpt(message) {
      case CodeRegex(code, _) ⇒ new MessageCode(code)
    }

  def apply(level: SchedulerLogLevel, line: String): LogEvent =
    level match {
      case SchedulerLogLevel.info ⇒ InfoLogged(line)
      case SchedulerLogLevel.warning ⇒ WarningLogged(line)
      case SchedulerLogLevel.error ⇒ ErrorLogged(line)
      case _ ⇒ OtherLevelLogged(level, line)
    }

  implicit object LogEventJsonFormat extends SimpleTypedJsonFormat[LogEvent] {
    protected def typeField = TypedJsonFormat.DefaultTypeFieldName → JsString("Logged")

    protected val subclasses = Set[Class[_ <: LogEvent]](
      classOf[InfoLogged],
      classOf[WarningLogged],
      classOf[ErrorLogged],
      classOf[OtherLevelLogged])

    private implicit def levelJsonFormat = SchedulerLogLevel.MyJsonFormat
    private val transferJsonFormat = jsonFormat2(OtherLevelLogged)  // Here, we (mis)use OtherLevelLogged as a transfer class

    protected def typelessWrite(o: LogEvent) =
      transferJsonFormat.write(OtherLevelLogged(o.level, o.message)).asJsObject

    def read(jsValue: JsValue) = {
      val OtherLevelLogged(level, message) = transferJsonFormat.read(jsValue)
      LogEvent(level, message)
    }
  }
}
